File: exercises/exercises_solutions.R
#Lecture 1:
rain<-read.csv("rainfall_summer_2014_simple.csv")
#check what you got:
head(rain)
#days appear to be in order, so to draw the barplots...
barplot(rain$rainfall)


#Lecture 2:
#This part was in the lecture notes:
API_BASE_URL<-"http://example.org/api/"

response<-getURL(apiurl)
response<-fromJSON(response)
min_round<-response[["/api/round/<round>"]]$param_min
max_round<-response[["/api/round/<round>"]]$param_max

#and here's the rest of it:
rounds<-min_round:max_round
nrounds<-length(rounds)

response<-fromJSON(getURL(paste0(apiurl,"round/",min_round)))
response$result

results<-matrix(NA,ncol=7,nrow=nrounds)
for(i in 1:nrounds) {  
  result<-fromJSON(getURL(paste0(apiurl,"round/",rounds[i])))$result
  #the following if is needed because some rounds are missing
  if(length(result)==7) results[i,]<-result
}
#we could format this as a data frame but if we are only interested
#in the number frequencies, we can just use it like a vector:
numbers<-as.vector(results)
barplot(table(numbers))

#also, reading from the JSON file:
lotto2000<-fromJSON("lotto-2000.json")
#a cumbersome list... to format this as an array you have to do 
#something like this
nrounds<-length(lotto2000)
numbers<-matrix(NA,ncol=7,nrow=nrounds)
for(i in 1:nrounds) numbers[i,]<-lotto2000[[i]][[2]]
#a more R-like way to do the same:
numbers<-sapply(lotto2000,FUN=function(x) x[[2]])
#and still to a vector etc.
numbers<-as.vector(results)
barplot(table(numbers))


#Lecture 3:
custdata<-read.table("https://github.com/WinVector/zmPDSwR/raw/master/Custdata/custdata.tsv",
                      header=TRUE,sep="\t")
summary(custdata)
#you should immdediatelly see that some incomes are negative
#and that the range of ages is 0 to 146.7
#Is that how it should be?
hist(custdata$income)
#there are some (maybe 1?) very high incomes
hist(custdata$age)
#the ages above 100 are probably some kind of an error?
table(custdata$age)
#also, things to think about: what does it mean that is.employed is missing?
#are the 56 missing values in housing.type, recent.move and num.vehicles the
#same people? how would you find out?

#next, broken slurm data
slurm<-read.table("broken_slurm.csv",sep="|",header=TRUE)
#"line 16 did not have 6 elements"
slurm<-read.table("broken_slurm.csv",sep="|",header=TRUE,fill=TRUE)
summary(slurm)
#Why is there a 6th variable that is completely empty?
#Also, the repeating 20 missing values should raise suspicion
#But it seems there are no missing values in User
summary(slurm$User) 
#something is really fishy! 
#to read the troublesome 16th line
scan("broken_slurm.csv",skip=15,nlines=1,what="character")
#ok, we are starting to see the problem
scan("broken_slurm.csv",skip=15,nlines=2,what="character")
#the lines are broken! We'll get back to fixing this later

#Lecture 4:
custdata2<-read.table("https://github.com/WinVector/zmPDSwR/raw/master/Custdata/custdata2.tsv",
                     header=TRUE,sep="\t")
#recode the TRUE, FALSE and NA values to "employed", "not employed" and "missing"
custdata2$is.employed.fix<-ifelse(custdata2$is.employed,"employed","not employed")
#NA's will stay NA's through that
custdata2$is.employed.fix<-ifelse(is.na(custdata2$is.employed.fix),"missing",custdata2$is.employed.fix)
custdata2$is.employed.fix<-factor(custdata2$is.employed.fix)
#you could also use the fact that booleans become 1 and 0 if you cast them as numeric and start from there!

custdata2$income.group<-cut(custdata2$income,breaks=c(0,10000,50000,100000,250000,1000000),
                            include.lowest=TRUE)
summary(custdata2$income.group)


#Lecture 5:
roo<-read.csv("roots_small.csv")
roowide<-reshape(roo,v.names="length",timevar="meas",idvar="root",direction="wide")
roowide2<-cbind(roowide[1],roowide[3:16]-roowide[2:15])

roolong<-reshape(roowide2,varying=2:15,v.names="growth",timevar="meas",times=2:15,direction="long")
rooagg<-aggregate(growth~meas,roolong,FUN=max,na.rm=TRUE)
plot(rooagg$growth,type="l")

File: exercises/exercises_solutions.py
import pandas as pd
import matplotlib.pyplot as pl

weather = pd.read_csv(
    "../datasets/weather-kumpula.csv",
    header=0,
    index_col=0,
    parse_dates=True)

weather.head()
weather['rmm'].plot()


ax1 = weather.t.plot(style='r-', legend=False, title="Weather 2014 (Kumpula, Helsinki)")
ax2 = weather.rmm.plot(secondary_y=True, style='b.')
ax1.set_ylabel("average temperature per day (celcius)")
ax2.right_ax.set_ylabel("cumulated rainfall per day (mm)")


#Lecture 2:
#This part was in the lecture notes:
API_BASE_URL<-"http://example.org/api/"
response = requests.get(API_BASE_URL).json()
min_round = response['/api/round/<round>']['param_min']
max_round = response['/api/round/<round>']['param_max']
print "min_round: %s max_round: %s" % (min_round, max_round)

results = []
for i in range(min_round, max_round + 1):
    response = requests.get(API_BASE_URL + 'round/%s' % (min_round + i)).json()
    result = response.get('result')
    if result:
        results.extend(result)

numbers = pd.Series(results)
numbers.hist(bins=39, normed=True)

# also, reading from the JSON file:
import json
lotto2000 = json.load(open("../datasets/lotto-2000.json"))
results = []
for l in lotto2000:
    results.extend(l[1])


#Lecture 3:
custdata = pd.read_table("https://github.com/WinVector/zmPDSwR/raw/master/Custdata/custdata.tsv")
custdata.describe(include='all')

#you should immdediatelly see that some incomes are negative
#and that the range of ages is 0 to 146.7
#Is that how it should be?
custdata.income.hist()
#there are some (maybe 1?) very high incomes
custdata.age.hist()
#the ages above 100 are probably some kind of an error?
sorted(custdata.age.unique())
#also, things to think about: what does it mean that is.employed is missing?
#are the 56 missing values in housing.type, recent.move and num.vehicles the
#same people? how would you find out?

#next, broken slurm data
slurm = pd.read_table("../datasets/broken_slurm.csv", sep="|",)
slurm.describe(include='all')

#column Unnamed: 0...
#Why is there a 6th variable that is completely empty?
#Also, the repeating 20 missing values should raise suspicion
#But it seems there are no missing values in User
slurm.User.unique()

#something is really fishy! 
slurm.ix[13:16, ]
#the lines are broken! We'll get back to fixing this later

#Lecture 4:
custdata2<-read.table("https://github.com/WinVector/zmPDSwR/raw/master/Custdata/custdata2.tsv",
                     header=TRUE,sep="\t")
mappings = {True: 'employed', False: 'not employed', np.NaN: 'missing'}
custdata2['is.employed'] = custdata2['is.employed'].map(mappings)

custdata2['income.group'] = pd.cut(custdata2.income, [0,10000,50000,100000,250000,1000000], include_lowest=True)

custdata2['income.group'].describe()


#Lecture 5:
roo<-read.csv("../datasets/roots_small.csv")
roowide = roo.pivot(index='root', columns='meas')


# Because of data alignment, the column names must match
#s1 = roowide.ix[:,:-1]
#s2 = roowide.ix[:,1:]
#s1 = s1.rename(columns= dict([(x, x+1) for x in range(1, 14)]))
#roowide2 = s2 - s1

# a bit cumbersome...

# Column renaming can be avoided with numpy, but data alignment is lost
#roowide2 = DataFrame(np.array(roowide.ix[:,1:]) - np.array(roowide.ix[:,:-1]))
#roowide2.index = roowide.index


roowide2 = pd.np.subtract(roowide.ix[:,1:], roowide.ix[:,:-1])

roolong = roowide2.stack().reset_index()

rooagg = roolong.groupby('meas').max()
rooagg.plot()
